Fix configuration cache incompatibility in Java Gradle Plugin Plugin

mlopatkin opened the pull request:
Plugin validation action relied on sharing List and Set between the validating action and file handlers. This doesn't work well with the configuration cache serialization, as it treats collections as values without identity. As a result of deserialization, each Action got a different copy of the collection, so the mutations weren't visible to the validation action.

The fix wraps the collections into a class. The class identity is preserved when deserializing, so all Actions share the same instance of the state class.

Fixes #22545

mlopatkin commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"e9009e9def0cadd555bc67d378108fda33d1a6ed","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
I've triggered the following builds for you:
- [PullRequestFeedback build](https://builds.gradle.org/viewQueued.html?itemId=67621387)

abstratt commented on subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java:
```diff
@@ -282,23 +280,62 @@ private Provider<JavaLauncher> toolchainLauncher(Project project) {
+     * The state is first collected by the {@link Jar#filesMatching(String, Action)} during the bundling.
+     * Then {@link PluginValidationAction} (added to the Jar as doLast) processes the collected information to emit warnings if necessary.
+     * <p>
+     * List and Set cannot be shared without this wrapper class as their identity isn't preserved by the configuration cache.
```
This makes sense as a description of your design for this change, but not as much as explaining the resulting code.

Instead, I would say use more general terms like wrapping the state into this class is required so it remains a shared object when storing it into/loading it from the configuration cache (collections in are duplicated when serialized).

abstratt reviewed (approved):
This kind of gotcha should be documented somewhere in the configuration cache documentation (so people don't make the same "mistake" in their own plugins). 

Otherwise, just a suggestion for changing a comment to be less about this change/bug and more about the design rationale per se.

abstratt commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"e9009e9def0cadd555bc67d378108fda33d1a6ed","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
I've triggered the following builds for you:
- [PullRequestFeedback build](https://builds.gradle.org/viewQueued.html?itemId=67626974)

mlopatkin commented:
@bot-gradle test PRF, BD

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"b5cea7d9266ffebee7eddc6a73655dd8f1e6f2df","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
I've triggered the following builds for you:
- [PullRequestFeedback build](https://builds.gradle.org/viewQueued.html?itemId=67767223)
- [BuildDistributions build](https://builds.gradle.org/viewQueued.html?itemId=67767097)

mlopatkin commented:
> This kind of gotcha should be documented somewhere in the configuration cache documentation (so people don't make the same "mistake" in their own plugins).

Good idea, I've added documentation.

> Otherwise, just a suggestion for changing a comment to be less about this change/bug and more about the design rationale per se.

Cleaned up the Javadoc, so it shouldn't be too elaborate now.

mlopatkin commented:
[Link to the changed docs](https://builds.gradle.org/repository/download/Gradle_Master_Check_BuildDistributions/67767097:id/distributions/gradle-8.3-docs.zip!/gradle-8.3-20230628110610%2B0000/docs/userguide/configuration_cache.html#config_cache:requirements:shared_objects)

lkasso commented on subprojects/docs/src/docs/userguide/running-builds/configuration_cache.adoc:
```diff
@@ -837,6 +837,43 @@ Instead, tasks should be connected using <<lazy_configuration#working_with_task_
+[[config_cache:requirements:shared_objects]]
+=== Sharing mutable objects
+When storing a task to the configuration cache, all objects directly or indirectly referenced through the task's fields are serialized.
+In most cases, deserialization preserves reference equality: if two fields `a` and `b` referenced the same instance at configuration time, then upon deserialization they will reference the same instance again, so `a == b` (or `a === b` in Groovy and Kotlin syntax) still holds.
```
```suggestion
In most cases, deserialization preserves reference equality: if two fields `a` and `b` reference the same instance at configuration time, then upon deserialization they will reference the same instance again, so `a == b` (or `a === b` in Groovy and Kotlin syntax) still holds.
```

lkasso commented on subprojects/docs/src/docs/userguide/running-builds/configuration_cache.adoc:
```diff
@@ -837,6 +837,43 @@ Instead, tasks should be connected using <<lazy_configuration#working_with_task_
+<3> Compare the reference to `ArrayList` instance stored in the task field and the reference captured in the `doLast` action.
+<4> Check the equality of stored and captured lists.
+
+Running the build without configuration cache shows that reference equality is preserved in both cases.
```
```suggestion
Running the build without the configuration cache shows that reference equality is preserved in both cases.
```

lkasso commented on subprojects/docs/src/docs/userguide/running-builds/configuration_cache.adoc:
```diff
@@ -837,6 +837,43 @@ Instead, tasks should be connected using <<lazy_configuration#working_with_task_
+
+In general, it isn't recommended to share mutable objects between configuration and execution phases.
+If you need to do this, you should always wrap the state in a class you define.
+There's no guarantee that the reference equality is preserved for standard Java, Groovy, and Kotlin types, or for Gradle-defined types.
```
```suggestion
There is no guarantee that the reference equality is preserved for standard Java, Groovy, and Kotlin types, or for Gradle-defined types.
```

lkasso reviewed (approved):
Very minor changes in docs.

mlopatkin commented:
@bot-gradle test and merge

TWiStErRob commented:
Really nice docs! Thanks for recording the gotcha!
